File: tests/machine/x/zig/group_by_multi_join.zig
const std = @import("std");

fn _sum_int(v: []const i32) i32 {
    var sum: i32 = 0;
    for (v) |it| { sum += it; }
    return sum;
}

fn _equal(a: anytype, b: anytype) bool {
    if (@TypeOf(a) != @TypeOf(b)) return false;
    return switch (@typeInfo(@TypeOf(a))) {
        .Struct, .Union, .Array, .Vector, .Pointer, .Slice => std.meta.eql(a, b),
        else => a == b,
    };
}

const NationsItem = struct {
    id: i32,
    name: []const u8,
};
const nations = &[_]NationsItem{
    NationsItem{
    .id = 1,
    .name = "A",
},
    NationsItem{
    .id = 2,
    .name = "B",
},
};
const SuppliersItem = struct {
    id: i32,
    nation: i32,
};
const suppliers = &[_]SuppliersItem{
    SuppliersItem{
    .id = 1,
    .nation = 1,
},
    SuppliersItem{
    .id = 2,
    .nation = 2,
},
};
const PartsuppItem = struct {
    part: i32,
    supplier: i32,
    cost: f64,
    qty: i32,
};
const partsupp = &[_]PartsuppItem{
    PartsuppItem{
    .part = 100,
    .supplier = 1,
    .cost = 10.0,
    .qty = 2,
},
    PartsuppItem{
    .part = 100,
    .supplier = 2,
    .cost = 20.0,
    .qty = 1,
},
    PartsuppItem{
    .part = 200,
    .supplier = 1,
    .cost = 5.0,
    .qty = 3,
},
};
const filtered = blk0: { var _tmp0 = std.ArrayList(struct {
    part: i32,
    value: f64,
}).init(std.heap.page_allocator); for (partsupp) |ps| { for (suppliers) |s| { if (!((s.id == ps.supplier))) continue; for (nations) |n| { if (!((n.id == s.nation))) continue; if (!(std.mem.eql(u8, n.name, "A"))) continue; _tmp0.append(struct {
    part: i32,
    value: f64,
}{
    .part = ps.part,
    .value = (ps.cost * ps.qty),
}) catch unreachable; } } } const _tmp1 = _tmp0.toOwnedSlice() catch unreachable; break :blk0 _tmp1; };
const grouped = blk2: { var _tmp4 = std.ArrayList(struct { key: i32, Items: std.ArrayList(std.StringHashMap(i32)) }).init(std.heap.page_allocator); var _tmp5 = std.AutoHashMap(i32, usize).init(std.heap.page_allocator); for (filtered) |x| { const _tmp6 = x.part; if (_tmp5.get(_tmp6)) |idx| { _tmp4.items[idx].Items.append(x) catch unreachable; } else { var g = struct { key: i32, Items: std.ArrayList(std.StringHashMap(i32)) }{ .key = _tmp6, .Items = std.ArrayList(std.StringHashMap(i32)).init(std.heap.page_allocator) }; g.Items.append(x) catch unreachable; _tmp4.append(g) catch unreachable; _tmp5.put(_tmp6, _tmp4.items.len - 1) catch unreachable; } } var _tmp7 = std.ArrayList(struct { key: i32, Items: std.ArrayList(std.StringHashMap(i32)) }).init(std.heap.page_allocator);for (_tmp4.items) |g| { _tmp7.append(g) catch unreachable; } var _tmp8 = std.ArrayList(struct {
    part: i32,
    total: f64,
}).init(std.heap.page_allocator);for (_tmp7.items) |g| { _tmp8.append(struct {
    part: i32,
    total: f64,
}{
    .part = g.key,
    .total = _sum_int(blk1: { var _tmp2 = std.ArrayList(i32).init(std.heap.page_allocator); for (g) |r| { _tmp2.append(r.value) catch unreachable; } const _tmp3 = _tmp2.toOwnedSlice() catch unreachable; break :blk1 _tmp3; }),
}) catch unreachable; } const _tmp8Slice = _tmp8.toOwnedSlice() catch unreachable; break :blk2 _tmp8Slice; };

pub fn main() void {
    std.debug.print("{any}\n", .{grouped});
}

File: tests/machine/x/cpp/group_by_multi_join.cpp
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

struct Nation {
  decltype(1) id;
  decltype(std::string("A")) name;
};
inline bool operator==(const Nation &a, const Nation &b) {
  return a.id == b.id && a.name == b.name;
}
inline bool operator!=(const Nation &a, const Nation &b) { return !(a == b); }
struct Supplier {
  decltype(1) id;
  decltype(1) nation;
};
inline bool operator==(const Supplier &a, const Supplier &b) {
  return a.id == b.id && a.nation == b.nation;
}
inline bool operator!=(const Supplier &a, const Supplier &b) {
  return !(a == b);
}
struct Partsupp {
  decltype(100) part;
  decltype(1) supplier;
  decltype(10) cost;
  decltype(2) qty;
};
inline bool operator==(const Partsupp &a, const Partsupp &b) {
  return a.part == b.part && a.supplier == b.supplier && a.cost == b.cost &&
         a.qty == b.qty;
}
inline bool operator!=(const Partsupp &a, const Partsupp &b) {
  return !(a == b);
}
struct Filtered {
  decltype(std::declval<Partsupp>().part) part;
  decltype((std::declval<Partsupp>().cost *
            std::declval<Partsupp>().qty)) value;
};
inline bool operator==(const Filtered &a, const Filtered &b) {
  return a.part == b.part && a.value == b.value;
}
inline bool operator!=(const Filtered &a, const Filtered &b) {
  return !(a == b);
}
struct __struct5 {
  decltype(std::declval<Filtered>().part) key;
  std::vector<Filtered> items;
};
inline bool operator==(const __struct5 &a, const __struct5 &b) {
  return a.key == b.key && a.items == b.items;
}
inline bool operator!=(const __struct5 &a, const __struct5 &b) {
  return !(a == b);
}
struct Grouped {
  decltype(std::declval<__struct5>().key) part;
  double total;
};
inline bool operator==(const Grouped &a, const Grouped &b) {
  return a.part == b.part && a.total == b.total;
}
inline bool operator!=(const Grouped &a, const Grouped &b) { return !(a == b); }
int main() {
  std::vector<Nation> nations = std::vector<Nation>{
      Nation{1, std::string("A")}, Nation{2, std::string("B")}};
  std::vector<Supplier> suppliers =
      std::vector<Supplier>{Supplier{1, 1}, Supplier{2, 2}};
  std::vector<Partsupp> partsupp = std::vector<Partsupp>{
      Partsupp{100, 1, 10, 2}, Partsupp{100, 2, 20, 1}, Partsupp{200, 1, 5, 3}};
  auto filtered = ([&]() {
    std::vector<Filtered> __items;
    for (auto ps : partsupp) {
      for (auto s : suppliers) {
        if (!((s.id == ps.supplier)))
          continue;
        for (auto n : nations) {
          if (!((n.id == s.nation)))
            continue;
          if (!((n.name == std::string("A"))))
            continue;
          __items.push_back(Filtered{ps.part, (ps.cost * ps.qty)});
        }
      }
    }
    return __items;
  })();
  auto grouped = ([&]() {
    std::vector<__struct5> __groups;
    for (auto x : filtered) {
      auto __key = x.part;
      bool __found = false;
      for (auto &__g : __groups) {
        if (__g.key == __key) {
          __g.items.push_back(Filtered{x});
          __found = true;
          break;
        }
      }
      if (!__found) {
        __groups.push_back(
            __struct5{__key, std::vector<Filtered>{Filtered{x}}});
      }
    }
    std::vector<Grouped> __items;
    for (auto &g : __groups) {
      __items.push_back(Grouped{
          g.key, ([&](auto v) {
            return std::accumulate(v.begin(), v.end(), 0.0);
          })(([&]() {
            std::vector<decltype(std::declval<Filtered>().value)> __items;
            for (auto r : g.items) {
              __items.push_back(r.value);
            }
            return __items;
          })())});
    }
    return __items;
  })();
  {
    auto __tmp1 = grouped;
    bool first = true;
    for (const auto &_x : __tmp1) {
      if (!first)
        std::cout << ' ';
      first = false;
      std::cout << "<struct>";
    }
    std::cout << std::endl;
  }
  return 0;
}
